flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorage.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.filesystem;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.CheckpointStorageLocation;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * An implementation of durable checkpoint storage to file systems.
+ */
+public class FsCheckpointStorage extends AbstractFsCheckpointStorage {
+
+	private final FileSystem fileSystem;
+
+	private final Path checkpointsDirectory;
+
+	private final Path sharedStateDirectory;
+
+	private final Path taskOwnedStateDirectory;
```

Comment:
As discussed, we might not need this if we map task-owned state to shared state. Idea: as long as the task still requires the state, it will resend the handle to this file (a placeholder to be precise) and keep the shared registry's reference count above zero.

Reply:
We may be able to remove this in the future, but because we need to pull task owned state out of the "exclusive state" directory soon, and the rework to make this use shared state handles is probably more delicate, I would actually have this for now. Should be trivial rework to remove it once we don't need it any more.